File: lib/rate-limit.ts
import { Redis } from '@upstash/redis';

const WINDOW_SECONDS = 60;
const LIMIT = 20;

const memory = new Map<string, { count: number; resetAt: number }>();

function getRedis(): Redis | null {
  if (!process.env.UPSTASH_REDIS_REST_URL || !process.env.UPSTASH_REDIS_REST_TOKEN) return null;
  return new Redis({
    url: process.env.UPSTASH_REDIS_REST_URL,
    token: process.env.UPSTASH_REDIS_REST_TOKEN,
  });
}

export async function checkRateLimit(identifier: string): Promise<{ allowed: boolean; retryAfter: number }> {
  const redis = getRedis();
  if (redis) {
    const key = `webshield:scan:${identifier}`;
    const count = await redis.incr(key);
    if (count === 1) await redis.expire(key, WINDOW_SECONDS);
    return {
      allowed: count <= LIMIT,
      retryAfter: count > LIMIT ? WINDOW_SECONDS : 0,
    };
  }

  if (process.env.NODE_ENV === 'production') {
    throw new Error('Distributed rate limiting is not configured.');
  }

  const now = Date.now();
  const current = memory.get(identifier);
  if (!current || current.resetAt <= now) {
    memory.set(identifier, { count: 1, resetAt: now + WINDOW_SECONDS * 1000 });
    return { allowed: true, retryAfter: 0 };
  }

  current.count += 1;
  return {
    allowed: current.count <= LIMIT,
    retryAfter: current.count > LIMIT ? Math.ceil((current.resetAt - now) / 1000) : 0,
  };
}

File: backend/rate_limit.py
import os
import time
from threading import Lock

try:
    from upstash_redis import Redis
except ImportError:
    Redis = None

LIMIT = 20
WINDOW_SECONDS = 60
_memory = {}
_lock = Lock()
_redis = None

if Redis and os.getenv("UPSTASH_REDIS_REST_URL") and os.getenv("UPSTASH_REDIS_REST_TOKEN"):
    _redis = Redis(
        url=os.environ["UPSTASH_REDIS_REST_URL"],
        token=os.environ["UPSTASH_REDIS_REST_TOKEN"],
    )


def check_rate_limit(identifier: str) -> tuple[bool, int]:
    if _redis:
        key = f"webshield:backend:scan:{identifier}"
        count = int(_redis.incr(key))
        if count == 1:
            _redis.expire(key, WINDOW_SECONDS)
        return count <= LIMIT, WINDOW_SECONDS if count > LIMIT else 0

    if os.getenv("ENVIRONMENT", "development").lower() == "production":
        raise RuntimeError("Distributed rate limiting is not configured.")

    now = time.time()
    with _lock:
        current = _memory.get(identifier)
        if not current or current[1] <= now:
            _memory[identifier] = [1, now + WINDOW_SECONDS]
            return True, 0
        current[0] += 1
        return current[0] <= LIMIT, max(1, int(current[1] - now)) if current[0] > LIMIT else 0
